### ISSUE TITLE
[Serve] DeploymentResponse._to_object_ref() blocks untill final results from actor

### ISSUE THREAD

**Issue by kyle-v6x** [ray-project/ray#46893]
### What happened + What you expected to happen

Since `2.10.0`, DeploymentResponses._to_object_ref() functions await the final result from the task, rather than returning the ObjectRef of the running task once it is scheduled. This effectively prevents any Async task from being computed while previous tasks are running (i.e passing by ref).

[You can find more details here.](https://github.com/ray-project/ray/issues/40995#issuecomment-2244538264)

### Versions / Dependencies

This is an issue with `Ray[Serve]>=2.10.0`

### Reproduction script

```
import asyncio
from ray import serve
from ray.serve.handle import DeploymentHandle
import time
from fastapi import FastAPI

app = FastAPI()


async def _to_object_ref_or_gen(
        self,
        hold=False
):
    """
    This replicates the behaviour locally for easier debugging.
    """
    obj_ref_or_gen = await asyncio.wrap_future(self._object_ref_future)
    if hold:
        obj_ref_or_gen = await obj_ref_or_gen.__anext__() # This call blocks until the result
    else:
        obj_ref_or_gen = obj_ref_or_gen.__anext__()

    self._object_ref_or_gen = obj_ref_or_gen

    return self._object_ref_or_gen

@serve.deployment(
    ray_actor_options={"num_cpus": 0},
    num_replicas=1
)
@serve.ingress(app)
class Dispatcher:
    def __init__(self, foo_handler: DeploymentHandle):
        self.foo_handler = foo_handler

    @app.post("/")
    async def entry(self, hold: bool):
        handle = None
        metrics = {}
        try:
            start = time.time()
            handle = self.foo_handler.remote()
            metrics["call_time"], start = time.time() - start, time.time()
            ref = await asyncio.wait_for(_to_object_ref_or_gen(handle, hold), timeout=30)
            metrics["scheduling_time"], start = time.time() - start, time.time()
            result = await ref
            metrics["worker_time"] = time.time() - start
        except TimeoutError:
            if handle is not None:
                handle.cancel()
            raise TimeoutError("Scheduler timeout error. All workers seemingly full.")
        finally:
            print(f"\n\nMetrics: {metrics}\n\n")
        return result


@serve.deployment(
    ray_actor_options={"num_cpus": 0},
    num_replicas=1
)
class Foo:
    def __call__(self):
        time.sleep(10)
        return True


foo = Foo.bind()
service = Dispatcher.bind(foo_handler=foo)


if __name__ == "__main__":
    from ray.cluster_utils import Cluster

    cluster = Cluster(
        initialize_head=True,
        head_node_args={
            "num_cpus": 4,
            "num_gpus": 0,
            "resources": {"head": 1},
            "dashboard_host": "0.0.0.0",
        },
    )
    worker_node = cluster.add_node(
        num_cpus=4,
        num_gpus=2,
        resources={"worker": 1},
    )

    cluster.wait_for_nodes(2)
    deployment = serve.run(service)

    print("\n\nTesting programmatically...\n\n")
    deployment.entry.remote(hold=False).result()
    deployment.entry.remote(hold=True).result()

    input("\n\nTest script completed. Press Enter to shutdown.\n\n")
    serve.shutdown()
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by shrekris-anyscale** [comment]
I was able to reproduce the issue too:

```python
# File name: repro.py

import asyncio
import time


from ray import serve


@serve.deployment(max_ongoing_requests=5)
async def sleepy():
    await asyncio.sleep(3)
    return "Finished!"

app = sleepy.bind()

async def experiment():
    handle = serve.run(app)
    print("Starting experiment.")

    t0 = time.perf_counter()
    deployment_response = handle.remote()
    t1 = time.perf_counter()
    print(f"Finished remote call: {(t1 - t0):.4f}s")

    t2 = time.perf_counter()
    assignment_ref = deployment_response._to_object_ref()
    t3 = time.perf_counter()
    print(f"Created assignment ref: {(t3 - t2):.4f}s")

    t4 = time.perf_counter()
    result_ref = await assignment_ref
    t5 = time.perf_counter()
    print(f"Awaited assignment ref: {(t5 - t4):.4f}s")

    t6 = time.perf_counter()
    result = await result_ref
    t7 = time.perf_counter()
    print(f"Awaited result ref: {(t7 - t6):.4f}s")

    print(f"Final results: {result}")


if __name__ == "__main__":
    asyncio.run(experiment())
```

```
% python repro.py

...
Starting experiment.
Finished remote call: 0.0023s
Created assignment ref: 0.0000s
Awaited assignment ref: 3.0111s
Awaited result ref: 0.0005s
Final results: Finished!
```

**Comment by zcin** [comment]
> This effectively prevents any Async task from being computed while previous tasks are running (i.e passing by ref).

Hi @kyle-v6x can you say more about this, and what you are trying to use the `_to_object_ref()` API for?

**Comment by kyle-v6x** [comment]
@zcin In order to pass the result of a serve replica to an actor async, we need to retrieve the `ObjectRef` and send it to the actor, as the actor has no concept of a `DeploymentResponse`. [Essentially the use case explained in the documentation here.](https://docs.ray.io/en/latest/serve/model_composition.html#advanced-convert-a-deploymentresponse-to-a-ray-objectref)

Further, if you want to run a task fully async [(I know this is not supported officially yet)](https://github.com/ray-project/ray/issues/32292#issuecomment-1855142111) it makes sense to retrieve/return the `task_id` for tracking. Since we can't obtain the `ObjectRef` async, we can't get the `task_id`.

[I'll link this one more time for visibility.](https://github.com/ray-project/ray/issues/40995#issuecomment-2244538264) I may have time next week to finally build locally and work on a solution, but I can't guaruntee!

**Comment by zcin** [comment]
Hi @kyle-v6x, your use case makes sense to me, and you're correct that currently that isn't supported in Serve because the second object ref won't be retrieved until the request finishes. We are currently working on a fix!

**Comment by GeneDer** [comment]
fixed by https://github.com/ray-project/ray/pull/47209

**Comment by GeneDer** [comment]
Sorry, not resolved yet. This is depended on https://github.com/ray-project/ray/issues/46934

**Comment by kyle-v6x** [comment]
Any updates here? I can't stress enough how critical this issue is for anyone depending on `async` functionality. Essentially all of our deployments depend on this, and we're stuck on a version which is well over a year old, and missing vital improvements. @GeneDer @jjyao

**Comment by kyle-v6x** [comment]
Checking in here again. If needed I can try look into it myself with some guidance. @jjyao was there an issue with the [existing draft ](https://github.com/ray-project/ray/pull/49586)you were working on? (cc. @GeneDer @edoakes )

**Comment by jjyao** [comment]
Hi @kyle-v6x, 

We currently have a bunch of P0 issues to fix. After that we will have someone work on it immediately. Sorry for the delay.

**Comment by kyle-v6x** [comment]
@jjyao As this is also a p0 issue, is there any timeline?

**Comment by jjyao** [comment]
@kyle-v6x we hope in July, we can have someone to work on it. Will keep it as P0 to make sure its top on our list.

**Comment by Stack-Attack** [comment]
Checking in once more. Hopefully the [Async RFC](https://github.com/ray-project/ray/issues/54652#issuecomment-3084927074) will provide an alternative, but this is still a pretty massive blocker for us.

**Comment by jjyao** [comment]
@MengjinYan will start to work on this starting this week.

**Comment by Stack-Attack** [comment]
@jjyao @MengjinYan Checking in again. Unfortunately, the new Async inference support is going to need quite a lot of work to replace our use-case, so the priority of this has been bumped up again.

**Comment by Stack-Attack** [comment]
We need to upgrade our clusters' Ray version, but this is still blocking us without some significant refactoring. It's to the point that we need to consider taking on the extra refactoring and abandoning the documented functionality of `to_object_ref`. 

Can we get some confirmation if this is ever going to be fixed?

**Comment by suppagoddo** [comment]
Picking this thread up. Verified the regression is still live on master and have a scoped mitigation to propose before I do into implementation phase. 

### Current State:
The 2.10 regression @Stack-Attack identified in [#2244538264](https://github.com/ray-project/ray/issues/40995#issuecomment-2244538264) is unchanged:
1. `send_request_python` routes the call through `handle_request_with_rejection.options(num_returns="streaming")` [Link](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/serve/_private/request_router/replica_wrapper.py#L99-L104) returning an `ObjectRefGenerator` even for unary user APIs.
2. `ActorReplicaResult.to_object_ref_async()` pulls the second yield via `await self._obj_ref_gen.__anext__()` [Link](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/serve/_private/replica_result.py#L238-L270).
3. `ObjectRefGenerator._next_async` awaits the full object value to detect readiness. [Link](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/_private/object_ref_generator.py#L254-L265), the `TODO(swang): Avoid fetching the value.` @Stack-Attack pointed at has moved from `_raylet.pyx:511` to [line 260
  here](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/_private/object_ref_generator.py#L260), but is otherwise unchanged.

### PRs which supposed to address this at core are closed:
1. `await response._to_object_ref()` still blocks until the computation finishes. 
2. PR #49586 is auto-closed. 

### Question:
@jjyao @MengjinYa: Seems like all the previous PRs where auto closed due to inactivity. If this is not actively worked on, I a have a proposed mitigation plan to unblock @kyle-v6x here. 

### Proposed mitigation: opt-in flag for non-streaming deployments:

**Not a replacement for the core fix**. A scoped workaround that restores pre-2.10 `_to_object_ref()` behavior for deployments where the user explicitly accepts softer backpressure.

The plain unary path already exists, it's [the third branch in `send_request_python`](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/serve/_private/request_router/replica_wrapper.py#L111), which returns a normal `ObjectRef`. It's currently only reachable when `with_rejection=False`, which only happens for Java replicas and Ray-Client contexts.

Add a deployment option (working name `strict_max_ongoing_requests_rejection: bool = True`). When set to `False` on a unary deployment:
  - [Router computes `with_rejection=False`](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/serve/_private/router.py#L975-L978) for that deployment.
  - `send_request_python` falls through to [the plain `handle_request` branch](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/serve/_private/request_router/replica_wrapper.py#L111).
  - `.remote()` returns a regular `ObjectRef` → `_to_object_ref()` resolves at scheduling time → chaining works → the pre-2.10 `asyncio.wait_for(response._to_object_ref(), timeout=N)` pattern works again.

**Tradeoff:** in-flight rejection is skipped for that deployment. The router's queue-length cache becomes the only guard against overshooting `max_ongoing_requests`. Under heavy contention the cache can be momentarily stale and the replica may end up with slightly more in-flight requests than its declared max identical semantics to how Serve already handles Java replicas, so the failure mode is established.

**Default stays `True`.** Zero behavior change for existing deployments. Users opt in per deployment.

### Scope and non-goals

Fixes the chaining use case and the scheduling-timeout workaround for unary deployments. Does **not** touch streaming deployments, does **not** fix the `_next_async` TODO, does **not** help users who need both strict rejection *and* working `_to_object_ref()`.

**FYI - This would not close #46893.** The core fix still matters; this is a mitigation so users can upgrade and unblock chaining today.

  ### Implementation sketch

Same pattern as #62160 (`rolling_update_percentage`): new `DeploymentConfig` field + `DeploymentSchema` + proto + decorator parameter + a single change in [`router.py:975-978`](https://github.com/ray-project/ray/blob/6eacf1dd7a/python/ray/serve/_private/router.py#L975-L978) to consult the flag. Serve-only, no Core changes.

  ### Asks
  - @edoakes @GeneDer @abrarsheikh: any thoughts on design? 
  - @kyle-v6x:  would this unblock your upgrade and chaining use cases, or is there a case you'd still be stuck on?

  Happy to open a draft PR if everything looks okay.

**Comment by abrarsheikh** [comment]
@kyle-v6x if I remember correctly, your requests are long-running, where you kick off model fine-tuning on the replica, which means these requests are long-running eg: over 1m. If that is true, then violating max_ongoing_requests is a non-starter.

**Comment by suppagoddo** [comment]
@abrarsheikh - If this is a long running task the mitigation which I proposed above will not work, in that case the fix needs to happen the core level which is we need to start with this PR[currently closed]: https://github.com/ray-project/ray/pull/49586 ... I can resurrect this one [ rebase it and see if anything else is needed in this PR] .... Once this is merged then I need to use this PR and apply another fix on top of it..

**Comment by kyle-v6x** [comment]
Apologies for the delay, I was away for a couple of weeks.

As far as I'm aware, [strict in-flight rejection for max_ongoing_requests was only added in 2.10.0](https://github.com/ray-project/ray/pull/42943) since it depends on the streaming changes, so we're already running without such a hard guarantee. In our production deployments, the slight over-attribution to one replica is easily absorbed with little impact. In longer-running cases (e.g., fine-tuning), the pipeline calls many other deployments in parallel/series — we only need one _to_object_ref call at the beginning of the pipeline to branch the code.

@suppagoddo this does seem like a valid solution to our issue! As you mentioned, it's not ideal and doesn't fix the core problem, but it should let us upgrade safely and then migrate away from this pattern toward the new APIs.
In general the new APIs are more robust for async calls but come with added complexity (Redis, etc.), so I think it's still worth keeping a Ray-native async flow available for smaller projects and simpler request structures.

Let me know how if I can assist in any way.
